server/repository/src/db_diesel/document.rs

```diff
@@ -0,0 +1,322 @@
+            .filter(document_head::dsl::id.eq(make_head_id(store_id, document_name)))
+            .first(&self.connection.connection)
+            .optional()?;
+        Ok(result)
+    }
+
+    pub fn query_heads(
+        &self,
+        filter: Option<DocumentHeadFilter>,
+    ) -> Result<Vec<DocumentHeadRow>, RepositoryError> {
+        let mut query = document_head::dsl::document_head.into_boxed();
+        if let Some(f) = filter {
+            apply_equal_filter!(query, f.name, document_head::dsl::name);
+            apply_equal_filter!(query, f.store_id, document_head::dsl::store_id);
+        }
+        let result = query.load(&self.connection.connection)?;
+        Ok(result)
+    }
+
+    /// Gets ancestor details for the full document history.
+    pub fn ancestor_details(
+        &self,
+        document_name: &str,
+    ) -> Result<Vec<AncestorDetail>, RepositoryError> {
+        let rows: Vec<(String, String, NaiveDateTime)> = document::dsl::document
+            .filter(document::dsl::name.eq(document_name))
+            .select((
+                document::dsl::id,
+                document::dsl::parents,
+                document::dsl::timestamp,
+            ))
+            .load(&self.connection.connection)?;
+        let mut ancestors = Vec::<AncestorDetail>::new();
+        for row in rows {
+            let parents: Vec<String> =
+                serde_json::from_str(&row.1).map_err(|err| RepositoryError::DBError {
+                    msg: "Invalid parents data".to_string(),
+                    extra: format!("{}", err),
+                })?;
+            ancestors.push(AncestorDetail {
+                id: row.0,
+                parents,
+                timestamp: row.2,
+            })
+        }
+        Ok(ancestors)
+    }
+}
+
+fn document_from_row(row: DocumentRow) -> Result<Document, RepositoryError> {
+    let parents: Vec<String> =
+        serde_json::from_str(&row.parents).map_err(|err| RepositoryError::DBError {
+            msg: "Invalid parents data".to_string(),
+            extra: format!("{}", err),
+        })?;
+    let data: serde_json::Value =
+        serde_json::from_str(&row.data).map_err(|err| RepositoryError::DBError {
+            msg: "Invalid data".to_string(),
+            extra: format!("{}", err),
+        })?;
+
+    let document = Document {
+        id: row.id,
+        name: row.name,
+        parents,
+        author: row.author,
+        timestamp: DateTime::<Utc>::from_utc(row.timestamp, Utc),
+        r#type: row.r#type,
+        data,
+        schema_id: row.schema_id,
+    };
+
+    Ok(document)
+}
+
+fn row_from_document(doc: &Document) -> Result<DocumentRow, RepositoryError> {
+    let parents = serde_json::to_string(&doc.parents).map_err(|err| RepositoryError::DBError {
+        msg: "Can't serialize parents".to_string(),
+        extra: format!("{}", err),
+    })?;
+    let data = serde_json::to_string(&doc.data).map_err(|err| RepositoryError::DBError {
```

Comment:
I think db_serialise and db_deserialise would be useful there are few places where it's used (both return DBError and both accept error message as second argument)

Reply:
you mean for serde for doc.parents and doc.data?